File: gestalt-core/src/main/java/org/config/gestalt/Gestalt.java
package org.config.gestalt;

import org.config.gestalt.exceptions.GestaltException;
import org.config.gestalt.reflect.TypeCapture;

public interface Gestalt {
    void loadConfigs() throws GestaltException;

    <T> T getConfig(String path, Class<T> klass) throws GestaltException;

    <T> T getConfig(String path, TypeCapture<T> klass) throws GestaltException;

    <T> T getConfig(String path, Class<T> klass, T defaultVal);

    <T> T getConfig(String path, TypeCapture<T> klass, T defaultVal);
}

File: gestalt-kotlin/src/main/kotlin/org/github/gestalt/config/kotlin/Gestalt.kt
package org.github.gestalt.config.kotlin

import org.github.gestalt.config.Gestalt
import org.github.gestalt.config.builder.GestaltBuilder
import org.github.gestalt.config.kotlin.decoder.*
import org.github.gestalt.config.kotlin.reflect.KTypeCapture
import kotlin.reflect.typeOf

/**
 * reified function to get a config, that automatically gets the generic type.
 */
@OptIn(ExperimentalStdlibApi::class)
inline fun <reified T> Gestalt.getConfig(path: String): T = this.getConfig(path, KTypeCapture.of<T>(typeOf<T>())) as T

/**
 * reified function to get a config with a default, that automatically gets the generic type.
 */
@OptIn(ExperimentalStdlibApi::class)
inline fun <reified T> Gestalt.getConfig(path: String, default: T): T = this.getConfig(path, default, KTypeCapture.of<T>(typeOf<T>())) as T

/**
 * Extension function for the GestaltBuilder that adds add kotlin decoders, and the default java ones.
 */
fun GestaltBuilder.addDefaultDecodersAndKotlin(): GestaltBuilder {
    val decoders = listOf(
        BooleanDecoder(), ByteDecoder(), CharDecoder(), DataClassDecoder(), DoubleDecoder(), FloatDecoder(), IntegerDecoder(),
        LongDecoder(), ShortDecoder(), StringDecoder()
    )

    return this.addDecoders(decoders).addDefaultDecoders()
}